core/modules/v1/login/model/providers/azureADLoginProvider.cfc

```diff
@@ -0,0 +1,94 @@
+component extends="baseLoginProvider" accessors=true output=false {
+
+	variables.providerName='azuread';
+
+	public function generateAuthUrl(state) {
+
+		setReturnURL();
+
+		var authUrl = "https://login.microsoftonline.com/#variables.configBean.getAzureADTenantID()#/oauth2/v2.0/authorize?";
+		authUrl &= "client_id=#variables.configBean.getAzureADClientID()#&";
+		authUrl &= "response_type=code&";
+		authUrl &= "redirect_uri=#urlEncodedFormat(getCallbackURL())#&";
+		authUrl &= "response_mode=query&";
+		authUrl &= "scope=https%3A%2F%2Fgraph.microsoft.com%2Fuser.read&";
+		authUrl &= "state=#urlEncodedFormat(state)#&";
+
+		return authUrl;
+	}
+
+	/*
+	I handle validating the code result from Azure AD and automatically getting the auth token.
+	I should be able to handle any bad result from Azuez AD or the user not allowing crap.
+	I also validate the state.
+	*/
+	public struct function validateResult(code, error, remoteState, clientState) {
+		var result = {};
+
+		//If error is anything, we have an error
+		if(error != "") {
+			result.status = false;
+			result.message = error;
+			return result;
+		}
+
+		//Then, ensure states are equal
+		if(remoteState != clientState) {
+			result.status = false;
+			result.message = "State values did not match.";
+			return result;
+		}
+
+		var token = getAuthToken(code);
+
+		var profile = getProfile(token.access_token);
+
+		profile.username = len(profile.mail) ? profile.mail : profile.userPrincipalName;
+		profile.given_name = !isNull(profile.givenName) ? profile.givenName : '';
```

Comment:
Same as the previous issue: is it possible to create users in Masa CMS without a firstname and lastname. And of so, do we want that for new Azure AD users? Or should an error be thrown?

Reply:
It is possible programatically to create users without firstname and lastname. Off course that's not ideal. Since Azure AD allows it as well, we need this flexibility.